Migrate to slog

bufdev opened the pull request:
This uses the same `zapcore.Core`, so logging output is exactly the same for now - we rely on the log format for printing warnings in a certain way, and the log format as-is is nicer for CLI output anyways. We can look to change that out in the future.

mfridman commented on private/pkg/slogtest/slogtest.go:
```diff
@@ -0,0 +1,32 @@
+// limitations under the License.
+
+// Package slogtest provides Loggers for testing.
+package slogtest
```
Suggestion, we name this literally anything else, this conflicts with https://pkg.go.dev/testing/slogtest which trips up imports.

Maybe we could have just an `slogext` package as the one way within buf to extend (and work) with slog, e.g., 

slogext.NewTestLogger
slogext.NopLogger
slogext.NewBuilder

etc.

bufdev replied:
I'll rename to `slogtestext` for now. I have this long-term plan to rename all the `.*ext` stuff to `x.*` ie `xslices`, `xslog`, but haven't gotten there yet. By naming this `slogtestext`, it'll map to `xslogtest` in that world.

bufdev replied:
Done

mfridman commented on private/pkg/slogtest/slogtest.go:
```diff
@@ -0,0 +1,32 @@
+)
+
+// NewLogger returns a new Logger for testing.
+func NewLogger(t testing.TB) *slog.Logger {
```
Wdyt about adding functional options to tune logging in testing, specifically log level is one we use often.

https://pkg.go.dev/go.uber.org/zap/zaptest#LoggerOption

bufdev replied:
Sure, give me a second

bufdev replied:
Done

bufdev commented on private/pkg/slogtestext/slogtestext.go:
```diff
@@ -0,0 +1,64 @@
+// WithLevel specifies the Level to use for the Logger.
+//
+// The default is slog.LevelInfo.
+func WithLevel(level slog.Level) LoggerOption {
```
@mfridman should the default be `slog.LevelDebug` for tests?

bufdev replied:
I think the answer is yes. This is also what `zaptest` did.

mfridman replied:
Yeah, where it was too noisy we bumped it to warn.

bufdev commented:
Big refactor just pushed up:

- Renamed `slogbuild` to `slogapp` and `zapbuild` to `zapapp`, as this really matches the other package names with regards to their purpose.
- Added `appext.LogFormat` and `appext.LogLevel`, including parsing for them.
- Made `slogapp` and `zapapp` talk in terms of `appext.LogFormat` and `appext.LogLevel`, as opposed to doing their own parsing.
- Added `appext.BuilderWithLoggerProvider` to allow custom loggers to be provided.
- Made a default `LoggerProvider` not rely on anything but the stdlib, so `appext` does not depend on `slogapp/zapapp`.
- Called `appext.BuilderWithLoggerProvider(slogapp.NewLogger)` in relevant places in the CLI.

Core should use `appext.BuilderWithLoggerProvider` to create a new `*slog.Logger` based on the input `io.Writer, appext.LogLevel, appext.LogFormat`.

bufdev commented on private/pkg/slogapp/slogapp.go:
```diff
@@ -0,0 +1,43 @@
+
+// LoggerProvider is an appext.LoggerProvider for use with appext.BuilderWithLoggerProvider.
+func LoggerProvider(container appext.NameContainer, logLevel appext.LogLevel, logFormat appext.LogFormat) (*slog.Logger, error) {
+	core, err := zapapp.NewCore(container.Stderr(), logLevel, logFormat)
```
Minor cleanup: this should call NewLogger

bufdev replied:
Done
